core/src/main/java/com/uber/m3/tally/sanitizers/ScopeSanitizerBuilder.java

```
@@ -0,0 +1,89 @@
+ * default, the name, key and value sanitize functions returns all the input
+ * untouched. Custom name, key or value Sanitize functions or ValidCharacters
+ * can be provided to override their default behaviour.
+ */
+public class ScopeSanitizerBuilder {
+
+    private Function<String, String> nameSanitizer = value -> value;
+    private Function<String, String> keySanitizer = value -> value;
+    private Function<String, String> valueSanitizer = value -> value;
+
+    private char repChar = ValidCharacters.DEFAULT_REPLACEMENT_CHARACTER;
+    private ValidCharacters nameCharacters;
+    private ValidCharacters keyCharacters;
+    private ValidCharacters valueCharacters;
+
+    public ScopeSanitizerBuilder withNameSanitizer(Function<String, String> nameSanitizer) {
+        this.nameSanitizer = nameSanitizer;
+        return this;
+    }
+
+    public ScopeSanitizerBuilder withKeySanitizer(Function<String, String> keySanitizer) {
+        this.keySanitizer = keySanitizer;
+        return this;
+    }
+
+    public ScopeSanitizerBuilder withValueSanitizer(Function<String, String> valueSanitizer) {
+        this.valueSanitizer = valueSanitizer;
+        return this;
+    }
+
+    public ScopeSanitizerBuilder withReplacementCharacter(char repChar) {
+        this.repChar = repChar;
+        return this;
+    }
+
+    public ScopeSanitizerBuilder withNameCharacters(ValidCharacters nameCharacters) {
+        this.nameCharacters = nameCharacters;
+        return this;
+    }
+
+    public ScopeSanitizerBuilder withKeyCharacters(ValidCharacters keyCharacters) {
+        this.keyCharacters = keyCharacters;
+        return this;
+    }
+
+    public ScopeSanitizerBuilder withValueCharacters(ValidCharacters valueCharacters) {
+        this.valueCharacters = valueCharacters;
+        return this;
+    }
+
+    public ScopeSanitizer build() {
+        if (nameCharacters != null) {
+            nameSanitizer = nameCharacters.sanitizeStringFunc(repChar);
```

if both `nameCharcters` and `nameSanitizer` provided (same for tag's keys and values) this will override the sanitizer and use sanitize function from `ValidCharacters`. If this is intentional behavior I think it should be documented more explicit. Or maybe throw an exception if someone tries to provide a sanitizer function when characters are already set?

---

good catch! I like throwing exception. Less document is better.

---

also added new tests for them.